src/Plutarch/Extra/Precompile.hs

```diff
@@ -0,0 +1,240 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+{-# HLINT ignore "Use camelCase" #-}
+
+{- | Pre-compiling Plutarch functions and applying them.
+
+ Speeds up benchmarking and testing.
+-}
+module Plutarch.Extra.Precompile (
+    applyScript,
+    -- Exporting the data constructor on purpose, since users might want to
+    -- deserialize compiled terms.  If someone wants to subvert type safety using
+    -- Scripts, they can do that regardless of this export.
+    CompiledTerm (..),
+    compile',
+    toDScript,
+    applyCompiledTerm,
+    applyCompiledTerm',
+    applyCompiledTerm2,
+    applyCompiledTerm2',
+    (##),
+    (##~),
+    (###),
+    (###~),
+    LiftError (..),
+    pliftCompiled',
+    pliftCompiled,
+) where
+
+import Control.Lens ((^?))
+import Data.Text (Text)
+import qualified Data.Text as Text
+import GHC.Stack (HasCallStack)
+import Plutarch.Evaluate (EvalError)
+import Plutarch.Extra.DScript (
+    DScript (DScript),
+    debugScript,
+    finalEvalDScript,
+    mustCompileD,
+    mustEvalD,
+    script,
+ )
+import Plutarch.Lift (
+    PConstantDecl (pconstantFromRepr),
+    PUnsafeLiftDecl (PLifted),
+ )
+import Plutarch.Prelude (PLift, S, Term, Type, (:-->))
+import PlutusCore.Builtin (KnownTypeError, readKnownConstant)
+import PlutusCore.Evaluation.Machine.Exception (_UnliftingErrorE)
+import PlutusLedgerApi.V1.Scripts (Script (Script, unScript))
+import UntypedPlutusCore (Program (Program, _progAnn, _progTerm, _progVer))
+import qualified UntypedPlutusCore.Core.Type as UplcType
+
+-- | Apply a function to an argument on the compiled 'Script' level.
+applyScript :: Script -> Script -> Script
+applyScript f a =
+    if fVer /= aVer
+        then error "apply: Plutus Core version mismatch"
+        else
+            Script
+                Program
+                    { _progTerm = UplcType.Apply () fTerm aTerm
+                    , _progVer = fVer
+                    , _progAnn = ()
+                    }
+  where
+    (Script Program{_progTerm = fTerm, _progVer = fVer}) = f
+    (Script Program{_progTerm = aTerm, _progVer = aVer}) = a
+
+applyDScript :: DScript -> DScript -> DScript
+applyDScript f a =
+    DScript
+        { script = applyScript f.script a.script
+        , debugScript = applyScript f.debugScript a.debugScript
+        }
+
+-- | Type-safe wrapper for compiled Plutarch functions.
+newtype CompiledTerm (a :: S -> Type) = CompiledTerm DScript
```

Comment:
neat pick:
couldn't this be 
```hs
newtype CompiledTerm (a :: S -> Type) = CompiledTerm { toDScript :: DScript }
```

Reply:
I'm disagree with this. `toDScript` is not a fitting name for a field, it's just abusing automatic generation of field selector functions. This makes using record-dot syntax as well as named field puns dealing with the field awkward. IMHO field selector functions should be deprecated anyway, as soon as record-dot syntax is solid.
Also, field selector functions on polymorphic fields (doesn't apply here) can have a type that has the `forall` in a bad place, potentially causing a need for eta-expansion (this was a hidden problem until simplified subsumption landed in GHC). For example, here the `forall` would be better of in the outer scope:
```
ghci> newtype Foo = Foo { ct :: forall (s :: S). Term s PInteger }
ghci> :t ct
ct :: Foo -> forall (s :: S). Term s PInteger
```

Reply:
Oh, I has some problems I was unaware of. Thanks for bringing this on.